Public/Src/Engine/Dll/Distribution/Grpc/ServerInterceptor.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using BuildXL.Engine.Distribution.Grpc;
+using BuildXL.Engine.Tracing;
+using BuildXL.Utilities.Instrumentation.Common;
+using Grpc.Core;
+using Grpc.Core.Interceptors;
+
+namespace BuildXL.Engine.Distribution
+{
+    internal class ServerInterceptor : Interceptor
+    {
+        private readonly LoggingContext m_loggingContext;
+        private readonly string m_buildId;
+
+        public ServerInterceptor(LoggingContext loggingContext, string buildId)
+        {
+            m_loggingContext = loggingContext;
+            m_buildId = buildId;
+        }
+
+        private void InterceptCallContext(ServerCallContext context)
+        {
+            string sender = context.Host;
+            string traceId = string.Empty;
+            string method = context.Method;
+            string senderBuildId = string.Empty;
+
+            foreach (var kvp in context.RequestHeaders)
+            {
+                if (kvp.Key == GrpcConstants.TraceIdKey)
+                {
+                    traceId = new Guid(kvp.ValueBytes).ToString();
+                }
+                else if (kvp.Key == GrpcConstants.BuildIdKey)
+                {
+                    senderBuildId = kvp.Value;
+                }
+            }
+
+            if (!string.IsNullOrEmpty(senderBuildId) && senderBuildId != m_buildId)
+            {

```

Comment:
>{ [](start = 12, length = 1)

Probably a good idea to log before throwing exception so we make sure to get something logged on the server.